core/src/main/java/com/linecorp/armeria/client/limit/ConcurrencyLimit.java

```diff
@@ -0,0 +1,141 @@
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+public final class ConcurrencyLimit {
+    private static final long DEFAULT_TIMEOUT_MILLIS = 10000L;
+
+    private final Function<RequestContext, Boolean> policy;
+    private final int maxConcurrency;
+    private final long timeoutMillis;
+    private final AtomicInteger numActiveRequests = new AtomicInteger();
+
+    /**
+     * Creates a new instance to limit the concurrent number of active requests to {@code maxConcurrency}.
+     *
+     * @param policy the policy to apply on the incoming {@code RequestContext} if the request should be
+     *      limited or not.
+     * @param maxConcurrency the maximum number of concurrent active requests. {@code 0} to disable the limit.
+     * @param timeoutMillis the amount of time until this decorator fails the request if the request was not
+     *                delegated to the {@code delegate} before then.
+     */
+    ConcurrencyLimit(Function<RequestContext, Boolean> policy, int maxConcurrency, long timeoutMillis) {
+        this.policy = policy;
+        this.maxConcurrency = maxConcurrency;
+        this.timeoutMillis = timeoutMillis;
+    }
+
+    /**
+     * Returns the value of the {@code "maxConcurrency"}.
+     */
+    public int maxConcurrency() {
+        return maxConcurrency;
+    }
+
+    /**
+     * Returns the value of the {@code "timeout"} with the desired {@code "unit"}.
+     */
+    public long timeout(TimeUnit desiredUnit) {
+        return desiredUnit.convert(timeoutMillis, MILLISECONDS);
+    }
+
+    /**
+     * Returns the value of the {@code "timeout"} in milliseconds.
+     */
+    public long timeoutMillis() {
+        return timeoutMillis;
+    }
+
+    /**
+     * Checks if the concurrency control should be enforced for the given {@code requestContext}.
+     */
+    public Boolean shouldLimit(RequestContext requestContext) {
+        return maxConcurrency > 0 && policy.apply(requestContext);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final ConcurrencyLimit that = (ConcurrencyLimit) o;
+        return maxConcurrency == that.maxConcurrency && timeoutMillis == that.timeoutMillis && policy.equals(
+                that.policy);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(policy, maxConcurrency, timeoutMillis);
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                          .add("policy", policy)
+                          .add("maxConcurrency", maxConcurrency)
+                          .add("timeoutMillis", timeoutMillis)
+                          .toString();
+    }
+
+    AtomicInteger numActiveRequests() {
+        return numActiveRequests;
+    }
```

Comment:
Could we move this method to above of `equals()`? We conventionally place `equals()`, `hashCode()` and `toString()` to the bottom of a class.